src/main/java/algorithms/curated170/medium/TheEarliestMomentWhenEveryoneBecomeFriends.java

```diff
@@ -0,0 +1,70 @@
+package algorithms.curated170.medium;
+
+import java.util.PriorityQueue;
+
+public class TheEarliestMomentWhenEveryoneBecomeFriends {
+
+    class Solution {
+
+        int[] parent;
+        int[] child;
+
+        public int earliestAcq(int[][] logs, int n) {
+            parent = new int[n];
+            child = new int[n];
+
+            int id_A, id_B, parent_A, parent_B;
+
+            for (int i = 0; i < n; i++) {
+                parent[i] = i;
+                child[i] = 1;
+            }
+
+            PriorityQueue<Log> pq = new PriorityQueue<>((a, b) -> a.timestamp - b.timestamp);
+            for (int i = 0; i < logs.length; i++) {
+                pq.add(new Log(logs[i][0], logs[i][1], logs[i][2]));
+            }
+
+            Log log;
+
+            while (!pq.isEmpty()) {
+                log = pq.poll();
+                id_A = log.id_A;
+                id_B = log.id_B;
+                parent_A = findParent(id_A);
+                parent_B = findParent(id_B);
+                if (parent_A == parent_B) {
+                    continue;
+                } else {
+                    parent[parent_A] = parent_B;
+                    child[parent_B] += child[parent_A];
+                    if (child[parent_B] == n) {
+                        return log.timestamp;
+                    }
+                }
+            }
+
+            return -1;
+        }
+
+        int findParent(int id) {
+            while (parent[id] != id) {
+                id = parent[id];
+            }
+            return id;
+        }
```

Comment:
Is it possible to compress parents?
```suggestion
        int findParent(int id) {
            while (parent[id] != id) {
                parent[id] = parent[parent[id]];
                id = parent[id];
            }
            return id;
        }
```

Reply:
I have to check up on it again, but it looks that the first code already compresses the tree from the beginning.

Reply:
It shouldn't on the grounds that the parent of the children of a parent remain same when the parent of the parent is changed.
For example the pairs:     [ 1-0, 2-1, 3-2 ]  would result in the tree ( 3->2->1->0 (root-parent) )

Reply:
It shouldn't on the grounds that the parent of the children of a parent remain same when the parent of the parent is changed.
For example the pairs: [ 1-0, 2-1, 3-2, 0-4, 4-5, 5-6 ] would result in the tree ( (1,2,3)->0->4->5->6 (root-parent) )

Reply:
Compression in and of itself has a cost but it should render more efficient if such IDs are called perhaps hundreds of times.

Reply:
Oke, we accept the suggestion. Thanks for the feedback.

Reply:
Compression takes a little bit more.
But it is up to the data population used by test cases in LeetCode.

![image](https://user-images.githubusercontent.com/60903744/120504808-1dc39100-c3cd-11eb-9867-cad15de705a1.png)
![image](https://user-images.githubusercontent.com/60903744/120504964-3a5fc900-c3cd-11eb-970a-fbd38607fd2b.png)